libs/core/config/include/hpx/config.hpp

```diff
@@ -202,13 +202,13 @@
 ///////////////////////////////////////////////////////////////////////////////
 /// By default, enable minimal thread deadlock detection in debug builds only.
 #if !defined(HPX_HAVE_THREAD_MINIMAL_DEADLOCK_DETECTION)
-#  if defined(HPX_DEBUG)
+#  if defined(HPX_DEBUG) && !defined(NDEBUG)
```

Comment:
Are you sure this is a good idea? This means that everything that depends on these defines (`HPX_HAVE_THREAD_MINIMAL_DEADLOCK_DETECTION`, `HPX_HAVE_SPINLOCK_DEADLOCK_DETECTION`, etc.) will depend on both `HPX_DEBUG` (which we control, as much as we can) and `NDEBUG` which the user is free to set as they wish in their application. I'm worried this will lead to ABI problems.

If this is for "disable lock verification for RelWithDebInfo builds", then `HPX_DEBUG` _should_ not be defined with `RelWithDebInfo`. If it is, something else is likely missing/wrong.

Reply:
Yes, I agree, this is a tricky one. On one hand, we want for the asserts to be in place for RelWithDebInfo (`HPX_DEBUG` is defined), otoh, the options related to thread debugging might not be desired by default for this build type. We could simply remove the conditional `#define ...` here and enable it through the build system for `Debug` builds instead.

Reply:
Where is `HPX_DEBUG` defined for `RelWithDebInfo`?

Reply:
Uhh, you got me there. I was _assuming_ it was defined (I believe to remember it was defined at some point). But if it's not (as it appears to be the case), I will revert the related changes.